Stats v4 UI: add date bar to the top of the charts

jaclync opened the pull request:
Fixes #1107  
Based on https://github.com/woocommerce/woocommerce-ios/pull/1188 since both branches have diffs on the same xib

- [x] ⚠️ update PR base to `develop` before merging ⚠️

## Changes
- Created `StatsTimeRangeBarView` that contains a label and updates its label text based on view model
- Created `StatsTimeRangeBarViewModel` that sets its `timeRangeText` based on start/end dates, site time zone, and stats time range
  - with basic unit tests (had to do manual date formatter for localization)
- Added `StatsTimeRangeBarView` view and a border view to the vertical stack view in `StoreStatsV4PeriodViewController.xib`. Configured the border view, and updated the `StatsTimeRangeBarView` when it's ready to load order data

## Testing
Prerequisite: the site/store has WC admin extension activated
- Build from Xcode with a store that has WC admin
- In "My store", check the charts in all tabs --> the text under the time range tab should correspond to "today"/"this week"/"this month"/"this year" as in the design

## Example screenshots

today | this week | this month | this year
--- | --- | --- | ---
![Simulator Screen Shot - iPhone 5s - 2019-08-19 at 10 52 43](https://user-images.githubusercontent.com/1945542/63236242-9183c080-c26f-11e9-8716-e97ea3ab5728.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-19 at 10 52 45](https://user-images.githubusercontent.com/1945542/63236243-9183c080-c26f-11e9-8e6d-e6c9b0417a48.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-19 at 10 52 47](https://user-images.githubusercontent.com/1945542/63236244-9183c080-c26f-11e9-92c9-74be77d67468.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-19 at 10 52 59](https://user-images.githubusercontent.com/1945542/63236246-921c5700-c26f-11e9-8e06-41815c91ffb9.png)

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

jaclync committed:
Add `noRevenueView` that includes `noRevenueLabel` in xib, configure these views and add UI update functions for the no revenue views and chart axis.

jaclync committed:
Determine whether there is revenue for an interval, with a unified `revenueValue` for chart y axis.

jaclync committed:
Create `StatsTimeRangeBarView` to display a time range.

jaclync committed:
Move logic for time range text to `StatsTimeRangeBarViewModel`.

jaclync committed:
Add time range bar view and bottom border view to stats v4 xib.

jaclync committed:
Add basic tests for `StatsTimeRangeBarViewModel` (have to manually set localized date format since the app can be in any locale when running tests).

jaclync committed:
Separate the format localizable string to a variable to shorten the line.

kyleaparker reviewed (changes requested):
For the today view, can we add the day as well? for example: Tuesday, Jun 29

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift:
```diff
@@ -56,6 +56,8 @@ class StoreStatsV4PeriodViewController: UIViewController {
     @IBOutlet private weak var chartAccessibilityView: UIView!
     @IBOutlet private weak var noRevenueView: UIView!
     @IBOutlet private weak var noRevenueLabel: UILabel!
+    @IBOutlet private weak var timeRangeBarView: StatsTimeRangeBarView!
```
Since this is in the .xib file and not code-generated, it will need autolayout constraints defined. There are several missing constraints for different views:

<img width="814" alt="Screen Shot 2019-08-16 at 9 54 52 AM" src="https://user-images.githubusercontent.com/1062444/63176902-6cdbde80-c00c-11e9-8951-6a42a08901d1.png">

When transitioning from landscape to portrait, the layout shifts:
![Simulator Screen Shot - iPhone Xs - 2019-08-16 at 09 59 37](https://user-images.githubusercontent.com/1062444/63177014-a9a7d580-c00c-11e9-8c36-b27d33b0ea4e.png)

<img src="https://user-images.githubusercontent.com/1062444/63177059-bfb59600-c00c-11e9-998f-d712cd4b96dc.png" width = "350" />

jaclync replied:
> Since this is in the .xib file and not code-generated, it will need autolayout constraints defined. There are several missing constraints for different views:

I believe the errors about Auto Layout constraints are from the xib not knowing the content of the custom view class `StatsTimeRangeBarView` since this view is defined in the code, not in Interface Builder. Normally, a `UIStackView`'s arranged subviews shouldn't need to specify the position along the stack view axis or height, given each subview has their height defined within their own and if the stack view's `alignment` and `distribution` are set to `fill`. I looked online, it looks like [a way to calm the xib Auto Layout checker](https://stackoverflow.com/questions/40933819/y-constraint-missing-in-vertical-stackview) is to add a height constraint at a lower priority and remove at build time.

jaclync replied:
> When transitioning from landscape to portrait, the layout shifts:

jaclync replied:
> When transitioning from landscape to portrait, the layout shifts:

do you have steps to reproduce this issue? I tested on iPhone 5s simulator and iPhone 6s Plus device, and wasn't able to repro this scroll view offset issue with a bunch of rotations

mindgraffiti replied:
Sure. With an iPhone XS simulator, run the app. When the My store tab is loaded, use Cmd + ➞ to rotate the screen, so the notch is on the righthand side. Then Cmd + ↑ to return to portrait mode. 

Here's a [demo movie](https://cld.wthms.co/h1uReN)...the animated .gif was too large to post here. I pulled the latest changes and can still reproduce this.

mindgraffiti replied:
> it looks like a way to calm the xib Auto Layout checker is to add a height constraint at a lower priority and remove at build time

lol. Works like a charm. Thank you!

mindgraffiti reviewed (changes requested):
This is changing so fast! Font change and autolayout changes requested.

jaclync committed:
Merge branch 'develop' into issue/1107-time-range-date-bar

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift
#	WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.xib
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj

jaclync committed:
Update "today" time range date formatter to show full day of week (e.g. "Monday, Aug 19").

jaclync committed:
Update date bar font.

jaclync committed:
Add a height constraint for time range bar view with low priority to get around constraint issue in xib.

jaclync committed:
Remove temporary height constraint at build time.

jaclync committed:
Update test.

jaclync commented:
> For the today view, can we add the day as well? for example: Tuesday, Jun 29

@kyleaparker I updated the screenshots in PR description, and here are some screenshots for the "today" tab in two locales:

en | zh-Hant
--- | ---
![Simulator Screen Shot - iPhone 5s - 2019-08-19 at 09 45 38](https://user-images.githubusercontent.com/1945542/63236318-d7d91f80-c26f-11e9-9095-8607d64286ed.png) | ![Simulator Screen Shot - iPhone 5s - 2019-08-19 at 09 46 10](https://user-images.githubusercontent.com/1945542/63236319-d7d91f80-c26f-11e9-888f-307993ad57fc.png)

@mindgraffiti I updated the PR with updates based on your suggestions, and responded to your Auto Layout / rotation comment for another pass!

mindgraffiti reviewed (changes requested):
Fixed - missing Steps to Reproduce for the shifted stats I'm experiencing, and added a video link demo. It bugs me that the screenshots don't render the notch, so you can't see which way the phone was rotated. 

The other changes look good!

jaclync committed:
Add a contain view to respond to safe area insets instead of child view controller's view.

jaclync committed:
Disable automatic content inset adjustment for `ButtonBarPagerTabStripViewController`'s scroll view.

jaclync committed:
Remove hacky override that doesn't seem needed anymore.

jaclync commented:
@mindgraffiti thanks a lot for the steps to repro and the video! The issue is from the safe area insets on devices with a notch (so it happens on iPhone X series but not on other iPhones, which I tested with). I updated the PR to:
- disable automatic content inset adjustment for the top-level content scroll view: https://github.com/woocommerce/woocommerce-ios/pull/1189/commits/7d064b9f6806fa81f4c9a0a0413e66e576bef778 (otherwise, the scroll view automatically gets a content offset from the safe area changes and this breaks the paging behavior in `ButtonBarPagerTabStripViewController`)
- add a container view for the top-level content child view controller so that the container responds to the safe area insets instead of the child view controller's view: https://github.com/woocommerce/woocommerce-ios/pull/1189/commits/9c823c906c3e15907a5b27c7366b941edb6f5ae2
- (and remove a hack that doesn't seem necessary from my testing: https://github.com/woocommerce/woocommerce-ios/pull/1189/commits/5906153fd281172074e5ae9840a145cc8ae89207)

I tested on iPhone XS and 5s simulators, and iPhone 6s Plus device. could you try if it works on your end?

mindgraffiti reviewed (approved):
Tested with the Xs sim. Autolayout issues fixed 💯 


:shipit:
